extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSinkP.java

```diff
@@ -0,0 +1,354 @@
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.jet.kinesis.impl;
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.services.kinesis.AmazonKinesisAsync;
+import com.amazonaws.services.kinesis.model.ProvisionedThroughputExceededException;
+import com.amazonaws.services.kinesis.model.PutRecordsRequestEntry;
+import com.amazonaws.services.kinesis.model.PutRecordsResult;
+import com.amazonaws.services.kinesis.model.PutRecordsResultEntry;
+import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.core.Inbox;
+import com.hazelcast.jet.core.Outbox;
+import com.hazelcast.jet.core.Processor;
+import com.hazelcast.jet.core.Watermark;
+import com.hazelcast.logging.ILogger;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
+
+public class KinesisSinkP<T> implements Processor {
+
+    //todo; Each shard can support writes up to 1,000 records per second, up to a
+    // maximum data write total of 1 MiB per second. Right now we don't pre-check
+    // this, because we don't check which item goes to what shard, we just rely
+    // failure handling with exponential backoff. Would be complicated to improve
+    // on, but can we affort not to?
+
+    /**
+     * PutRecords requests are limited to 500 records.
+     */
+    private static final int MAX_RECORDS_IN_REQUEST = 500;
+
+    private static final long PAUSE_AFTER_FAILURE = 1000L; //todo: exponential backoff
+
+    /**
+     * Each record, when encoded as a byte array, is limited to 1M,
+     * including the partition key (Unicode String).
+     */
+    private static final int MAX_RECORD_SIZE_IN_BYTES = 1024 * 1024;
+
+    /**
+     * The maximum allowed size of all the records in a PutRecords request,
+     * including partition keys is 5M.
+     */
+    private static final int MAX_REQUEST_SIZE_IN_BYTES = 5 * 1024 * 1024;
+
+    /**
+     * The number of Unicode characters making up keys is limited to a maximum
+     * of 256.
+     */
+    private static final int MAX_UNICODE_CHARS_IN_KEY = 256;
+
+    private final AmazonKinesisAsync kinesis;
+    private final String stream;
+
+    private final Buffer<T> buffer;
+
+    private ILogger logger;
+    private KinesisHelper helper;
+
+    private State state = State.READY_TO_SEND;
+    private long nextSendTime;
```

Comment:
Should initialize to MIN_VALUE, it won't start sending until 1.1.1970 otherwise :-)
Especially if we switch to `nanoTime`.